// react-navi/dist/commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Link_1 = require("./Link");
exports.Link = Link_1.Link;
var NaviProvider_1 = require("./NaviProvider");
exports.NaviProvider = NaviProvider_1.NaviProvider;
var NotFoundBoundary_1 = require("./NotFoundBoundary");
exports.NotFoundBoundary = NotFoundBoundary_1.NotFoundBoundary;
var Router_1 = require("./Router");
exports.Router = Router_1.Router;
var NaviConsumer_1 = require("./NaviConsumer");
exports.useNavigation = NaviConsumer_1.useNavigation;
exports.NaviConsumer = NaviConsumer_1.NaviConsumer;
var History_1 = require("./History");
exports.useHistory = History_1.useHistory;
exports.History = History_1.History;
var LoadingRoute_1 = require("./LoadingRoute");
exports.useLoadingRoute = LoadingRoute_1.useLoadingRoute;
exports.LoadingRoute = LoadingRoute_1.LoadingRoute;
var CurrentRoute_1 = require("./CurrentRoute");
exports.useCurrentRoute = CurrentRoute_1.useCurrentRoute;
exports.CurrentRoute = CurrentRoute_1.CurrentRoute;
var View_1 = require("./View");
exports.View = View_1.View;
// Aliases for compatibility with older versions
var View_2 = require("./View");
exports.NavContent = View_2.View;
var Link_2 = require("./Link");
exports.NavLink = Link_2.Link;
var NotFoundBoundary_2 = require("./NotFoundBoundary");
exports.NavNotFoundBoundary = NotFoundBoundary_2.NotFoundBoundary;
var NaviProvider_2 = require("./NaviProvider");
exports.NavProvider = NaviProvider_2.NaviProvider;
var History_2 = require("./History");
exports.NavHistory = History_2.History;
var LoadingRoute_2 = require("./LoadingRoute");
exports.NavLoading = LoadingRoute_2.LoadingRoute;
var CurrentRoute_2 = require("./CurrentRoute");
exports.NavRoute = CurrentRoute_2.CurrentRoute;
// Deprecated
var NaviContext_1 = require("./NaviContext");
exports.NavContextProvider = NaviContext_1.NavContextProvider;
