// navi/dist/commonjs/MemoryNavigation.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var history_1 = require("history");
var Navigation_1 = require("./Navigation");
function createMemoryNavigation(options) {
    if (options.pages) {
        if (process.env.NODE_ENV !== 'production') {
            console.warn("Deprecation Warning: passing a \"pages\" option to \"createMemoryNavigation()\" will " +
                "no longer be supported from Navi 0.13. Use the \"routes\" option instead.");
        }
        options.routes = options.pages;
    }
    var url = options.url || (options.request && options.request.url);
    if (!url) {
        throw new Error("createMemoryNavigation() could not find a URL.");
    }
    var history = history_1.createMemoryHistory({
        // The initial entry is ignored, and replaced during the call
        // to navigate below.
        initialEntries: ['/'],
    });
    var navigation = new Navigation_1.Navigation({
        history: history,
        basename: options.basename,
        context: options.context,
        routes: options.routes,
        trailingSlash: options.trailingSlash,
    });
    navigation.navigate(__assign({}, options.request, { url: url, replace: true }));
    return navigation;
}
exports.createMemoryNavigation = createMemoryNavigation;
